toolchain/base/value_store.h

```diff
@@ -84,8 +85,62 @@ namespace Internal {
 // Used as a parent class for non-printable types. This is just for
 // std::conditional, not as an API.
 class ValueStoreNotPrintable {};
+
+// Provides YAML printing of a value.
+template <typename ValueT>
+inline auto PrintValue(llvm::raw_ostream& out, const ValueT& val) {
+  out << val;
+};
+template <>
+inline auto PrintValue(llvm::raw_ostream& out, const llvm::APInt& val) {
+  val.print(out, /*isSigned=*/false);
+};
+template <>
+inline auto PrintValue(llvm::raw_ostream& out, const llvm::StringRef& val) {
+  out << "\"" << llvm::yaml::escape(val) << "\"";
+};
+
 }  // namespace Internal
 
+// Provides YAML printing of a list. The first line indent applies even if there
+// is no label, in which case it applies to the first element.
+template <typename ValueT>
+inline auto PrintValueRange(
+    llvm::raw_ostream& out, llvm::iterator_range<const ValueT*> range,
+    std::optional<llvm::StringRef> label, int first_line_indent,
+    int later_indent, bool trailing_newline,
+    std::function<void(llvm::raw_ostream&, const ValueT& val)> print =
+        Internal::PrintValue<ValueT>) {
```

Comment:
Using a `std::function` here feels a bit heavy / inefficient. We probably don't care much about efficiency for our YAML printing in general, but it looks like it'd be straightforward to switch to `llvm::function_ref` here, so I think we should do so as a general good practice, in the same way that we're passing in a `llvm::StringRef` instead of a `std::string`.